File: src/main/java/com/fyp/studentinterfaceservice/services/interfaces/PositionService.java
package com.fyp.studentinterfaceservice.services.interfaces;

import com.fyp.studentinterfaceservice.exceptions.ProgradException;
import com.fyp.studentinterfaceservice.exceptions.UserNotFoundException;
import com.fyp.studentinterfaceservice.model.Application;
import com.fyp.studentinterfaceservice.model.ApplicationWrapper;
import com.fyp.studentinterfaceservice.model.Position;
import org.springframework.http.ResponseEntity;

import java.text.ParseException;
import java.util.List;
import java.util.Map;

public interface PositionService {

    List<Position> getAllPositions();

    Position getPosition(Long id);

    List<Position> searchPositions(String location, String keywords);

    List<Position> searchJobsApi(String location, String keywords);

    List<Position> getCompanyPositions(Long companyId);

    ResponseEntity<String> apply(Application application) throws ProgradException;

    List<Application> getApplicationsByEmail(String email);

    Position update(Position position);

    List<Position> getJobRecommendations(String email) throws UserNotFoundException;

    ApplicationWrapper applicationsStats(String email);

    Application updateApplication(Application application);

    Map<String, Double> searchSalaries(String keywords, String location) throws ParseException;
}

File: src/main/resources/student-interface-service/src/app/service/position.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {PositionModel} from "../model/position.model";
import {ApplicationModel} from "../model/application.model";

@Injectable({
  providedIn: 'root'
})
export class PositionService {

  constructor(private http: HttpClient) {
  }

  searchJobsApi(location:string, keywords: string): Observable<Array<PositionModel>> {
    return this.http.get<Array<PositionModel>>('http://localhost:8083/searchJobsApi/'+ location + '/' + keywords)
  }

  searchPositions(location:string, keywords: string): Observable<Array<PositionModel>> {
    return this.http.get<Array<PositionModel>>('http://localhost:8083/searchPositions/'+ location + '/' + keywords)
  }

  searchSalaries(keywords: string, location: string): Observable<any> {
    return this.http.get('http://localhost:8083/searchSalary/' + keywords + '/' + location);
  }

  getCompanyPositions(companyId: number): Observable<any> {
    return this.http.get<Array<PositionModel>>('http://localhost:8083/getCompanyPositions/'+ companyId)
  }

  getAllJobs(): Observable<Array<PositionModel>> {
    return this.http.get<Array<PositionModel>>('http://localhost:8083/getAllPositions/')
  }

  getRecommendedJobs(email: string): Observable<any> {
    return this.http.get('http://localhost:8083/positions/recommend',
      {params: {email: email}});
  }

  // getRecommendations(email: string): any {
  //   if(!!localStorage.getItem('recommendedJobs')) {
  //     this.http.get('http://localhost:8083/positions/recommend',
  //       {params: {email: email}}).subscribe(data => {
  //       localStorage.setItem('recommendedJobs', JSON.stringify(data));
  //       return data;
  //     });
  //   }
  //   else {
  //     return JSON.parse(localStorage.getItem('recommendedJobs'));
  //   }
  // }


  getJob(id: number): Observable<PositionModel> {
    return this.http.get<PositionModel>('http://localhost:8083/getPosition/' + id);
  }

  apply(application: ApplicationModel): Observable<any> {
    return this.http.post('http://localhost:8083/apply', application,
      {responseType: 'blob'});
  }

  updateJob(position: PositionModel): Observable<any>{
    return this.http.put('http://localhost:8083/positions/update', position);
  }

  getUserApplicationsStats(email: string): Observable<any> {
    return this.http.get('http://localhost:8083/applicationsStats',
      {params: {email: email}});
  }

  updateApplication(application: ApplicationModel): Observable<any> {
    return this.http.put('http://localhost:8083/application/update', application);
  }
}
